src/Security/Authentication/Negotiate/src/NegotiateHandler.cs

```diff
@@ -0,0 +1,334 @@
+            };
+            await Events.Authenticated(authenticatedContext);
+
+            if (authenticatedContext.Result != null)
+            {
+                return authenticatedContext.Result;
+            }
+
+            var ticket = new AuthenticationTicket(authenticatedContext.Principal, authenticatedContext.Properties, Scheme.Name);
+            return AuthenticateResult.Success(ticket);
+        }
+
+        /// <summary>
+        /// Issues a 401 WWW-Authenticate Negotiate challenge.
+        /// </summary>
+        /// <param name="properties"></param>
+        /// <returns></returns>
+        protected override async Task HandleChallengeAsync(AuthenticationProperties properties)
+        {
+            // We allow issuing a challenge from an HTTP/2 request. Browser clients will gracefully downgrade to HTTP/1.1.
+            // SocketHttpHandler will not downgrade (https://github.com/dotnet/corefx/issues/35195), but WinHttpHandler will.
+            var eventContext = new ChallengeContext(Context, Scheme, Options, properties);
+            await Events.Challenge(eventContext);
+            if (eventContext.Handled)
+            {
+                return;
+            }
+
+            Logger.LogDebug($"Challenged 401 Negotiate");
+            Response.StatusCode = StatusCodes.Status401Unauthorized;
+            Response.Headers.Append(HeaderNames.WWWAuthenticate, NegotiateVerb);
+        }
+
+        private AuthPersistence EstablishConnectionPersistence(IDictionary<object, object> items)
+        {
+            Debug.Assert(!items.ContainsKey(AuthPersistenceKey), "This should only be registered once per connection");
+            var persistence = new AuthPersistence();
+            RegisterForConnectionDispose(persistence);
+            items[AuthPersistenceKey] = persistence;
+            return persistence;
+        }
+
+        private IDictionary<object, object> GetConnectionItems()
+        {
+            var connectionItems = Context.Features.Get<IConnectionItemsFeature>()?.Items;
+            if (connectionItems == null)
+            {
+                throw new NotSupportedException($"Negotiate authentication requires a server that supports {nameof(IConnectionItemsFeature)} like Kestrel.");
+            }
+
+            return connectionItems;
+        }
+
+        private void RegisterForConnectionDispose(IDisposable authState)
+        {
+            var connectionCompleteFeature = Context.Features.Get<IConnectionCompleteFeature>();
+            if (connectionCompleteFeature == null)
+            {
+                throw new NotSupportedException($"Negotiate authentication requires a server that supports {nameof(IConnectionCompleteFeature)} like Kestrel.");
+            }
+            connectionCompleteFeature.OnCompleted(DisposeState, authState);
+        }
+
+        private static Task DisposeState(object state)
+        {
+            ((IDisposable)state).Dispose();
+            return Task.CompletedTask;
```

Comment:
Is it guaranteed that state will implement IDisposable?

Reply:
Yes, in this scenario state is always AuthPersistence.

Reply:
Is it worth adding a Debug.Assert?

Reply:
If this were public or internal maybe, but since it's only used privately from one place I'm not concerned about it. Worst case we'll get a type case exception that should show up in tests.